agent/php_wrapper.c

```diff
@@ -56,16 +56,16 @@ static void nr_php_wraprec_add_before_after_clean_callbacks(
   wraprec->special_instrumentation_clean = clean_callback;
 }
 
-nruserfn_t* nr_php_wrap_user_function_before_after_clean_with_transience(
+nruserfn_t* nr_php_wrap_user_function_before_after_clean_with_options(
```

Comment:
There exist:
* `nr_php_wrap_user_function` that creates wraprec with options valid for library and framework wraprecs: `NR_WRAPREC_NOT_TRANSIENT`, `NR_WRAPREC_CREATE_INSTRUMENTED_FUNCTION_METRIC` (the options are embedded in the `nr_php_wrap_user_function` implementation)
* `nr_php_wrap_generic_callable` that creates wraprec with options valid for callables: `NR_WRAPREC_IS_TRANSIENT`, `NR_WRAPREC_NO_INSTRUMENTED_FUNCTION_METRIC` (the options are embedded in the `nr_php_wrap_generic_callable`)
* `nr_php_wrap_user_function_with_options` that creates wraprec with arbitrary options which is only used by `nr_php_wrap_generic_callable` and `nr_php_wrap_user_function` which both have options set properly for their use cases.
* `nr_php_wrap_user_function_before_after_clean_with_options` that creates wraprec with arbitrary options - used for both library and framework wraprecs and callables and in each case the options had to be set 'by hand'.

However, all library and framework instrumentation needs wraprecs that are not transient and create supportability metrics (that's why `nr_php_wrap_user_function` has the options 'hardcoded'). Moreover only callables wraprecs need to be transient and not create supportability metrics (that's why `nr_php_wrap_generic_callable` has the options 'hardcoded'). Therefore, wouldn't it makes sense to have `nr_php_wrap_user_function_before_after_clean` that corresponds to `nr_php_wrap_user_function`, and has the options for wraprecs for library and framework instrumentation 'hardcoded'? It will indicate and highlight that wraprecs for library and framework instrumentation need to be non-transient and create supportability metrics, regardless if zend_execute hook or oapi is used to instrument user functions. Such API will help to avoid mistakes of using incorrect options for wraprecs when instrumenting library and framework user functions.

Reply:
I prefer explicit coding where options/parameters are specified and clear at every call, even when redundant. However, I recognize that the existence of `nr_php_wrap_user_function` begs the existence of `nr_php_wrap_user_function_before_after_clean`. added in [449afaa](https://github.com/newrelic/newrelic-php-agent/pull/730/commits/449afaacda9a9d734919944b3742d90d348eb3d9)

Reply:
Thank you!